model/project_parser_storage.go

```diff
@@ -77,3 +78,37 @@ func ParserProjectUpsertOne(ctx context.Context, settings *evergreen.Settings, m
 	defer ppStorage.Close(ctx)
 	return ppStorage.UpsertOne(ctx, pp)
 }
+
+// ParserProjectUpsertOneWithS3Fallback attempts to upsert the parser project
+// into persistent storage using the given storage method. If it fails due to
+// DB document size limitations, it will attempt to fall back to using S3
+// to store it. If it succeeds, this returns the actual project storage method
+// used to persist the parser project; otherwise, it returns the
+// originally-requested storage method.
+func ParserProjectUpsertOneWithS3Fallback(ctx context.Context, settings *evergreen.Settings, method evergreen.ParserProjectStorageMethod, pp *ParserProject) (evergreen.ParserProjectStorageMethod, error) {
+	err := ParserProjectUpsertOne(ctx, settings, method, pp)
+	if err == nil {
+		return method, nil
+	}
+	if method == evergreen.ProjectStorageMethodS3 {
+		return method, err
+	}
+
+	if !db.IsDocumentLimit(err) {
+		return method, err
+	}
+
+	flags, flagErr := evergreen.GetServiceFlags()
+	if flagErr != nil {
+		return method, errors.Wrap(flagErr, "getting service flags to check ability to fall back to S3")
+	}
+	if flags.ParserProjectS3StorageDisabled {
+		return method, err
+	}
+
+	if err := ParserProjectUpsertOne(ctx, settings, evergreen.ProjectStorageMethodS3, pp); err != nil {
+		return method, errors.Wrap(err, "upserting parser project into S3")
```

Comment:
We should either return plain `err` here, or return a wrapped error above on L94 and L98

Reply:
I wrapped the other errors so it's a little more obvious which one it chose.